File: remote-aircraft/firmware/production/autonomous_drone.c
/*
 * autonomous_drone.c — Production autonomous drone firmware skeleton (bare-metal C)
 * Complexity level:  PRODUCTION
 * Language:          C (MISRA-C:2023 inspired, no dynamic allocation)
 *
 * This file represents the *flight management layer* running on a high-integrity
 * flight computer (e.g., Cube Orange+ STM32H7 or custom ARM Cortex-M7 board).
 * It is NOT a complete ArduPilot replacement — it illustrates the architectural
 * patterns used in production flight software.
 *
 * Architecture:
 *   ┌──────────────────────────────────────────────────────────────────┐
 *   │  Sensor fusion (EKF stub)  │  Mission manager  │  Safety monitor │
 *   ├──────────────────────────────────────────────────────────────────┤
 *   │            Attitude controller (cascaded PID)                    │
 *   ├──────────────────────────────────────────────────────────────────┤
 *   │            Motor mixer  │  Actuator output (DShot/PWM)           │
 *   └──────────────────────────────────────────────────────────────────┘
 *
 * Key production patterns demonstrated:
 *   - Fixed-size state machines (no malloc)
 *   - Watchdog-compatible main loop
 *   - Redundant sensor voting
 *   - Arming / safety interlock
 *   - MAVLink encode + UART DMA-ready API
 *   - Battery failsafe
 *
 * Compile (ARM Cortex-M7, STM32H743):
 *   arm-none-eabi-gcc -mcpu=cortex-m7 -mfpu=fpv5-d16 -mfloat-abi=hard \
 *     -mthumb -O2 -std=c11 -Wall -Wextra -Wpedantic \
 *     -DSTM32H743xx \
 *     -o autonomous_drone.elf autonomous_drone.c
 */

#include <stdint.h>
#include <stdbool.h>
#include <string.h>
#include <math.h>

/* ============================================================
 * Platform types and compile-time assertions
 * ============================================================ */
typedef uint8_t  u8;
typedef uint16_t u16;
typedef uint32_t u32;
typedef int16_t  s16;
typedef int32_t  s32;
typedef float    f32;

#define ARRAY_LEN(a) (sizeof(a) / sizeof((a)[0]))

/* Compile-time assertion (works in C11) */
#define STATIC_ASSERT(cond, msg) _Static_assert(cond, msg)

STATIC_ASSERT(sizeof(f32) == 4, "float must be 32-bit");

/* ============================================================
 * Constants
 * ============================================================ */
#define NUM_MOTORS          6U     /* hexacopter */
#define NUM_IMU_INSTANCES   3U     /* triple-redundant IMU */
#define MAX_WAYPOINTS       50U
#define LOOP_HZ             400U
#define HEARTBEAT_PERIOD_MS 1000U
#define BATT_FAILSAFE_PCT   20.0f

/* Motor pulse limits */
#define PWM_MIN_US 1000U
#define PWM_MAX_US 2000U
#define PWM_MID_US 1500U

/* ============================================================
 * PID controller (no heap, no function pointers in inner loop)
 * ============================================================ */
typedef struct {
    f32 kp, ki, kd;
    f32 integral;
    f32 prev_error;
    f32 integral_limit;
} PidController;

static void pid_reset(PidController *const p) {
    p->integral   = 0.0f;
    p->prev_error = 0.0f;
}

static f32 pid_update(PidController *const p, f32 setpoint,
                       f32 measured, f32 dt) {
    f32 error = setpoint - measured;
    p->integral += error * dt;
    if (p->integral >  p->integral_limit) { p->integral =  p->integral_limit; }
    if (p->integral < -p->integral_limit) { p->integral = -p->integral_limit; }
    f32 deriv = (dt > 0.0f) ? ((error - p->prev_error) / dt) : 0.0f;
    p->prev_error = error;
    return (p->kp * error) + (p->ki * p->integral) + (p->kd * deriv);
}

/* ============================================================
 * IMU sample + redundancy voting
 * ============================================================ */
typedef struct { f32 gx, gy, gz; } ImuSample;

/** Median-of-three vote on a single axis (fault tolerance). */
static f32 vote_axis(f32 a, f32 b, f32 c) {
    if ((a <= b && b <= c) || (c <= b && b <= a)) { return b; }
    if ((b <= a && a <= c) || (c <= a && a <= b)) { return a; }
    return c;
}

static ImuSample vote_imu(const ImuSample sensors[NUM_IMU_INSTANCES]) {
    ImuSample out;
    out.gx = vote_axis(sensors[0].gx, sensors[1].gx, sensors[2].gx);
    out.gy = vote_axis(sensors[0].gy, sensors[1].gy, sensors[2].gy);
    out.gz = vote_axis(sensors[0].gz, sensors[1].gz, sensors[2].gz);
    return out;
}

/* ============================================================
 * Mission manager
 * ============================================================ */
typedef struct {
    f32 lat, lon, alt_m;
    f32 speed_ms;
    bool release_payload;
} Waypoint;

typedef enum {
    MISSION_IDLE = 0,
    MISSION_RUNNING,
    MISSION_COMPLETE,
    MISSION_ABORTED,
} MissionState;

typedef struct {
    Waypoint    waypoints[MAX_WAYPOINTS];
    u8          count;
    u8          current_idx;
    MissionState state;
} MissionManager;

static void mission_init(MissionManager *const m) {
    memset(m, 0, sizeof(*m));
    m->state = MISSION_IDLE;
}

static bool mission_add_waypoint(MissionManager *const m, Waypoint wp) {
    if (m->count >= MAX_WAYPOINTS) { return false; }
    m->waypoints[m->count++] = wp;
    return true;
}

static void mission_start(MissionManager *const m) {
    if (m->count > 0U) {
        m->current_idx = 0U;
        m->state       = MISSION_RUNNING;
    }
}

static void mission_advance(MissionManager *const m) {
    m->current_idx++;
    if (m->current_idx >= m->count) {
        m->state = MISSION_COMPLETE;
    }
}

static const Waypoint *mission_current_wp(const MissionManager *const m) {
    if (m->state == MISSION_RUNNING) {
        return &m->waypoints[m->current_idx];
    }
    return NULL;
}

/* ============================================================
 * Safety monitor
 * ============================================================ */
typedef enum {
    SAFE_OK       = 0,
    SAFE_LOW_BATT = 1,
    SAFE_GEOFENCE = 2,
    SAFE_LINK_LOSS = 3,
    SAFE_IMU_FAIL  = 4,
} SafetyFlag;

typedef struct {
    u32 flags;               /* bitmask of SafetyFlag */
    f32 battery_pct;
    u32 last_rc_ms;          /* ms timestamp of last RC packet */
    u32 uptime_ms;
} SafetyMonitor;

static void safety_init(SafetyMonitor *const s) {
    memset(s, 0, sizeof(*s));
    s->battery_pct = 100.0f;
}

static void safety_update(SafetyMonitor *const s, u32 now_ms) {
    s->uptime_ms = now_ms;

    /* Battery failsafe */
    if (s->battery_pct < BATT_FAILSAFE_PCT) {
        s->flags |= (1U << (u32)SAFE_LOW_BATT);
    }

    /* RC link loss: no RC packet for > 1000 ms */
    if ((now_ms - s->last_rc_ms) > 1000U) {
        s->flags |= (1U << (u32)SAFE_LINK_LOSS);
    }
}

static bool safety_rtl_required(const SafetyMonitor *const s) {
    return (s->flags & ((1U << (u32)SAFE_LOW_BATT) |
                        (1U << (u32)SAFE_LINK_LOSS) |
                        (1U << (u32)SAFE_GEOFENCE))) != 0U;
}

/* ============================================================
 * Hexacopter motor mixer (flat X6)
 * ============================================================ */
static void mix_hex_x(f32 thr, f32 roll, f32 pitch, f32 yaw,
                       u16 out[NUM_MOTORS]) {
    /* Flat-X6 geometry, angles 30°/90°/150°/210°/270°/330° */
    static const f32 ROLL_DIR[6]  = { 0.5f, 1.0f,  0.5f, -0.5f, -1.0f, -0.5f};
    static const f32 PITCH_DIR[6] = { 1.0f, 0.0f, -1.0f, -1.0f,  0.0f,  1.0f};
    static const f32 YAW_DIR[6]   = {-1.0f, 1.0f, -1.0f,  1.0f, -1.0f,  1.0f};

    for (u8 i = 0U; i < NUM_MOTORS; i++) {
        f32 m = thr
              + roll  * ROLL_DIR[i]
              + pitch * PITCH_DIR[i]
              + yaw   * YAW_DIR[i];
        s32 us = (s32)PWM_MIN_US + (s32)(m);
        if (us < (s32)PWM_MIN_US) { us = (s32)PWM_MIN_US; }
        if (us > (s32)PWM_MAX_US) { us = (s32)PWM_MAX_US; }
        out[i] = (u16)us;
    }
}

/* ============================================================
 * Motor output stub (replace with DMA-backed timer CCR writes)
 * ============================================================ */
static void motor_set_all(const u16 pulse_us[NUM_MOTORS]) {
    /* In real firmware: write each pulse_us[i] to TIMx->CCRi */
    (void)pulse_us;
}

/* ============================================================
 * MAVLink HEARTBEAT encoder
 * ============================================================ */
#define MAVLINK_STX    0xFEU
#define MAV_SYS_ID     1U
#define MAV_COMP_ID    1U
#define MSG_HEARTBEAT  0U
#define HB_CRC_EXTRA   50U

static u16 mavlink_crc_accumulate(u8 byte, u16 crc) {
    u8 tmp = byte ^ (u8)(crc & 0xFFU);
    tmp ^= (tmp << 4U);
    return (u16)((crc >> 8U)
                 ^ ((u16)tmp << 8U)
                 ^ ((u16)tmp << 3U)
                 ^ ((u16)tmp >> 4U));
}

static u8 mavlink_seq = 0U;

/** Encode HEARTBEAT into caller-supplied buffer (min 17 bytes).
 *  Returns number of bytes written. */
static u8 encode_heartbeat(u8 *buf, u8 buf_len) {
    if (buf_len < 17U) { return 0U; }

    u8 payload[9] = {0};
    payload[4] = 6U;  /* MAV_TYPE_HEXAROTOR */
    payload[5] = 3U;  /* MAV_AUTOPILOT_ARDUPILOTMEGA */
    payload[8] = 3U;  /* MAVLink version 3 */

    u8 header[5] = {
        (u8)sizeof(payload),
        mavlink_seq++,
        MAV_SYS_ID,
        MAV_COMP_ID,
        MSG_HEARTBEAT
    };

    /* Compute CRC */
    u16 crc = 0xFFFFU;
    for (u8 i = 0U; i < (u8)sizeof(header); i++) {
        crc = mavlink_crc_accumulate(header[i], crc);
    }
    for (u8 i = 0U; i < (u8)sizeof(payload); i++) {
        crc = mavlink_crc_accumulate(payload[i], crc);
    }
    crc = mavlink_crc_accumulate(HB_CRC_EXTRA, crc);

    /* Assemble packet */
    u8 idx = 0U;
    buf[idx++] = MAVLINK_STX;
    for (u8 i = 0U; i < (u8)sizeof(header);  i++) { buf[idx++] = header[i];  }
    for (u8 i = 0U; i < (u8)sizeof(payload); i++) { buf[idx++] = payload[i]; }
    buf[idx++] = (u8)(crc & 0xFFU);
    buf[idx++] = (u8)(crc >> 8U);
    return idx;
}

/* ============================================================
 * SysTick (1 ms) — stub; replace with MCU vector
 * ============================================================ */
static volatile u32 g_tick_ms = 0U;
void SysTick_Handler(void) { g_tick_ms++; }

static void delay_ms(u32 ms) {
    u32 end = g_tick_ms + ms;
    while (g_tick_ms < end) { /* wait */ }
}

/* ============================================================
 * UART transmit stub
 * ============================================================ */
static void uart_write(const u8 *data, u8 len) {
    /* Replace with DMA UART write */
    (void)data; (void)len;
}

/* ============================================================
 * Main
 * ============================================================ */
int main(void) {
    /* ---- Hardware init (stubbed) ---- */
    /* systick_init();  i2c_init();  uart_init();  timer_init(); */

    /* ESC arming */
    u16 arm_pulse[NUM_MOTORS];
    for (u8 i = 0U; i < NUM_MOTORS; i++) { arm_pulse[i] = PWM_MIN_US; }
    motor_set_all(arm_pulse);
    delay_ms(2000U);

    /* ---- PID init ---- */
    PidController roll_pid  = {.kp=1.2f, .ki=0.03f, .kd=0.15f, .integral_limit=400.f};
    PidController pitch_pid = {.kp=1.2f, .ki=0.03f, .kd=0.15f, .integral_limit=400.f};
    PidController yaw_pid   = {.kp=0.8f, .ki=0.01f, .kd=0.00f, .integral_limit=200.f};
    pid_reset(&roll_pid); pid_reset(&pitch_pid); pid_reset(&yaw_pid);

    /* ---- Mission setup ---- */
    MissionManager mission;
    mission_init(&mission);
    mission_add_waypoint(&mission, (Waypoint){.lat=47.39f,.lon=8.54f,.alt_m=30.f,.speed_ms=12.f});
    mission_add_waypoint(&mission, (Waypoint){.lat=47.40f,.lon=8.54f,.alt_m=30.f,.speed_ms=12.f,.release_payload=true});
    mission_add_waypoint(&mission, (Waypoint){.lat=47.41f,.lon=8.54f,.alt_m=30.f,.speed_ms=12.f});
    mission_start(&mission);

    /* ---- Safety init ---- */
    SafetyMonitor safety;
    safety_init(&safety);

    /* ---- Control loop ---- */
    const f32 dt = 1.0f / (f32)LOOP_HZ;
    u32 hb_timer = 0U;
    u8  hb_buf[20];

    while (1) {
        u32 now = g_tick_ms;

        /* Read IMU (stub: all zeros) */
        ImuSample raw_imu[NUM_IMU_INSTANCES] = {{0}};
        ImuSample imu = vote_imu(raw_imu);

        /* RC setpoints (stub: level hover) */
        f32 rc_thr = 0.0f, rc_roll = 0.0f, rc_pitch = 0.0f, rc_yaw = 0.0f;
        safety.last_rc_ms = now;          /* update in real ISR */

        /* Safety check */
        safety_update(&safety, now);
        if (safety_rtl_required(&safety)) {
            /* Override setpoints to RTL behaviour */
            rc_thr  = 500.0f;
            rc_roll = rc_pitch = rc_yaw = 0.0f;
        }

        /* Attitude PID */
        f32 roll_out  = pid_update(&roll_pid,  rc_roll,  imu.gx, dt);
        f32 pitch_out = pid_update(&pitch_pid, rc_pitch, imu.gy, dt);
        f32 yaw_out   = pid_update(&yaw_pid,   rc_yaw,   imu.gz, dt);

        /* Motor mixer */
        u16 motor_us[NUM_MOTORS];
        mix_hex_x(rc_thr, roll_out, pitch_out, yaw_out, motor_us);
        motor_set_all(motor_us);

        /* Mission waypoint advance (simplified — replace with distance check) */
        if (mission.state == MISSION_RUNNING) {
            const Waypoint *wp = mission_current_wp(&mission);
            if (wp != NULL && wp->release_payload) {
                /* trigger payload release GPIO */
            }
            /* In a real system: advance when within acceptance_radius of wp */
            (void)wp;
        }

        /* Heartbeat at 1 Hz */
        if ((now - hb_timer) >= HEARTBEAT_PERIOD_MS) {
            u8 len = encode_heartbeat(hb_buf, (u8)sizeof(hb_buf));
            uart_write(hb_buf, len);
            hb_timer = now;
        }

        /* Maintain loop rate (watchdog pet goes here) */
        while ((g_tick_ms - now) < (1000U / LOOP_HZ));
    }
    /* never reached */
    return 0;
}

File: remote-aircraft/firmware/production/autonomous_drone.zig
//! autonomous_drone.zig — Production autonomous drone firmware skeleton (Zig)
//! Complexity level:  PRODUCTION
//! Language:          Zig 0.13+
//!
//! Architecture (all in a single file for portability):
//!   - Comptime-configurable subsystem parameters
//!   - Triple-redundant IMU voting (no allocation)
//!   - Hexacopter motor mixer
//!   - Mission manager (fixed-size waypoint array)
//!   - Safety monitor (battery + link-loss flags)
//!   - MAVLink HEARTBEAT encoder
//!   - Zero-heap attitude PID controller
//!
//! Build (STM32H743 Cortex-M7 with FPU):
//!   zig build-exe autonomous_drone.zig \
//!       -target thumb-freestanding-eabihf \
//!       -mcpu cortex_m7+vfp4 \
//!       -O ReleaseSafe \
//!       --name autonomous_drone
//!
//! Build (RP2040 Cortex-M0+ — no FPU):
//!   zig build-exe autonomous_drone.zig \
//!       -target thumb-freestanding-none \
//!       -mcpu cortex_m0plus \
//!       -O ReleaseSmall \
//!       --name autonomous_drone

const std = @import("std");

// ---------------------------------------------------------------------------
// Compile-time configuration
// ---------------------------------------------------------------------------

const Config = struct {
    num_motors:        comptime_int = 6,
    num_imu_instances: comptime_int = 3,
    max_waypoints:     comptime_int = 50,
    loop_hz:           comptime_int = 400,
    heartbeat_period:  comptime_int = 400, // ticks at loop_hz → 1 Hz
    batt_failsafe_pct: comptime_float = 20.0,
    pwm_min_us:        u32 = 1_000,
    pwm_max_us:        u32 = 2_000,
    integral_limit:    comptime_float = 400.0,
};

const CFG = Config{};
const DT: f32 = 1.0 / @as(f32, @floatFromInt(CFG.loop_hz));

// ---------------------------------------------------------------------------
// PID controller (zero allocation, comptime gains allowed)
// ---------------------------------------------------------------------------

const PidGains = struct { kp: f32, ki: f32, kd: f32 };

const PidState = struct {
    integral:   f32 = 0.0,
    prev_error: f32 = 0.0,

    pub fn reset(self: *PidState) void {
        self.integral   = 0.0;
        self.prev_error = 0.0;
    }

    pub fn update(self: *PidState, g: PidGains,
                  setpoint: f32, measured: f32, dt: f32) f32 {
        const err = setpoint - measured;
        self.integral = std.math.clamp(
            self.integral + err * dt,
            -CFG.integral_limit, CFG.integral_limit,
        );
        const deriv: f32 = if (dt > 0) (err - self.prev_error) / dt else 0.0;
        self.prev_error = err;
        return g.kp * err + g.ki * self.integral + g.kd * deriv;
    }
};

// ---------------------------------------------------------------------------
// IMU + redundancy voting
// ---------------------------------------------------------------------------

const ImuSample = struct { gx: f32, gy: f32, gz: f32 };

fn median3(a: f32, b: f32, c: f32) f32 {
    if ((a <= b and b <= c) or (c <= b and b <= a)) return b;
    if ((b <= a and a <= c) or (c <= a and a <= b)) return a;
    return c;
}

fn voteIMU(s: [CFG.num_imu_instances]ImuSample) ImuSample {
    return .{
        .gx = median3(s[0].gx, s[1].gx, s[2].gx),
        .gy = median3(s[0].gy, s[1].gy, s[2].gy),
        .gz = median3(s[0].gz, s[1].gz, s[2].gz),
    };
}

/// Read one IMU instance (stub — replace with I2C reads).
fn readIMUInstance(_: u8) ImuSample {
    return .{ .gx = 0, .gy = 0, .gz = 0 };
}

// ---------------------------------------------------------------------------
// RC input (stub)
// ---------------------------------------------------------------------------

const RcInput = struct {
    throttle: f32 = 0.0,
    roll:     f32 = 0.0,
    pitch:    f32 = 0.0,
    yaw:      f32 = 0.0,
    valid:    bool = true,
};

fn readRC() RcInput { return .{}; }

// ---------------------------------------------------------------------------
// Mission manager
// ---------------------------------------------------------------------------

const Waypoint = struct {
    lat:            f32,
    lon:            f32,
    alt_m:          f32,
    speed_ms:       f32 = 10.0,
    release_payload: bool = false,
};

const MissionStatus = enum { Idle, Running, Complete, Aborted };

const MissionManager = struct {
    waypoints:   [CFG.max_waypoints]Waypoint = undefined,
    count:       usize = 0,
    current_idx: usize = 0,
    status:      MissionStatus = .Idle,

    pub fn addWaypoint(self: *MissionManager, wp: Waypoint) bool {
        if (self.count >= CFG.max_waypoints) return false;
        self.waypoints[self.count] = wp;
        self.count += 1;
        return true;
    }

    pub fn start(self: *MissionManager) void {
        if (self.count > 0) {
            self.current_idx = 0;
            self.status      = .Running;
        }
    }

    pub fn currentWaypoint(self: *const MissionManager) ?Waypoint {
        if (self.status == .Running and self.current_idx < self.count) {
            return self.waypoints[self.current_idx];
        }
        return null;
    }

    pub fn advance(self: *MissionManager) void {
        self.current_idx += 1;
        if (self.current_idx >= self.count) self.status = .Complete;
    }
};

// ---------------------------------------------------------------------------
// Safety monitor
// ---------------------------------------------------------------------------

const SafetyFlags = packed struct(u32) {
    low_battery: bool = false,
    link_loss:   bool = false,
    geofence:    bool = false,
    imu_fail:    bool = false,
    _padding:    u28  = 0,
};

const SafetyMonitor = struct {
    flags:       SafetyFlags = .{},
    battery_pct: f32 = 100.0,
    rc_age_ticks: u32 = 0,   // incremented each loop tick; reset on RC update

    pub fn tick(self: *SafetyMonitor, rc: RcInput) void {
        if (rc.valid) {
            self.rc_age_ticks = 0;
        } else {
            self.rc_age_ticks += 1;
        }
        // 400 ticks = 1 second link loss
        if (self.rc_age_ticks > @as(u32, @intCast(CFG.loop_hz))) {
            self.flags.link_loss = true;
        }
        if (self.battery_pct < CFG.batt_failsafe_pct) {
            self.flags.low_battery = true;
        }
    }

    pub fn needsRTL(self: *const SafetyMonitor) bool {
        return self.flags.low_battery or self.flags.link_loss or self.flags.geofence;
    }
};

// ---------------------------------------------------------------------------
// Hex-X motor mixer (comptime geometry)
// ---------------------------------------------------------------------------

const HEX_ROLL  = [CFG.num_motors]f32{ 0.5,  1.0,  0.5, -0.5, -1.0, -0.5};
const HEX_PITCH = [CFG.num_motors]f32{ 1.0,  0.0, -1.0, -1.0,  0.0,  1.0};
const HEX_YAW   = [CFG.num_motors]f32{-1.0,  1.0, -1.0,  1.0, -1.0,  1.0};

fn mixHexX(thr: f32, roll: f32, pitch: f32, yaw: f32) [CFG.num_motors]u32 {
    var out: [CFG.num_motors]u32 = undefined;
    for (0..CFG.num_motors) |i| {
        const v = thr + roll * HEX_ROLL[i] + pitch * HEX_PITCH[i] + yaw * HEX_YAW[i];
        out[i] = std.math.clamp(
            @as(u32, @intFromFloat(1000.0 + v)),
            CFG.pwm_min_us, CFG.pwm_max_us,
        );
    }
    return out;
}

// ---------------------------------------------------------------------------
// Motor output (register stubs — replace with TIM CCR writes)
// ---------------------------------------------------------------------------

const MOTOR_REGS = [CFG.num_motors]u32{
    0x40000034, 0x40000038, 0x4000003C,
    0x40000040, 0x40000044, 0x40000048,
};

fn motorSet(idx: usize, pulse_us: u32) void {
    const reg = @as(*volatile u32, @ptrFromInt(MOTOR_REGS[idx]));
    reg.* = std.math.clamp(pulse_us, CFG.pwm_min_us, CFG.pwm_max_us);
}

fn motorSetAll(pulse_us: u32) void {
    for (0..CFG.num_motors) |i| motorSet(i, pulse_us);
}

// ---------------------------------------------------------------------------
// MAVLink HEARTBEAT encoder (no allocator)
// ---------------------------------------------------------------------------

const MAVLINK_STX: u8 = 0xFE;

fn mavCRC(data: []const u8) u16 {
    var crc: u16 = 0xFFFF;
    for (data) |b| {
        var tmp: u8 = b ^ @as(u8, @truncate(crc & 0xFF));
        tmp ^= tmp << 4;
        crc = (crc >> 8) ^
              (@as(u16, tmp) << 8) ^
              (@as(u16, tmp) << 3) ^
              (@as(u16, tmp) >> 4);
    }
    return crc;
}

fn encodeHeartbeat(seq: u8, buf: *[17]u8) void {
    // payload
    var p = [9]u8{0, 0, 0, 0, 13, 3, 0, 4, 3}; // HEXAROTOR, ArduPilot
    const hdr = [5]u8{ @as(u8, @intCast(p.len)), seq, 1, 1, 0 };
    // CRC
    var crc_input: [15]u8 = undefined;
    @memcpy(crc_input[0..5],  &hdr);
    @memcpy(crc_input[5..14], &p);
    crc_input[14] = 50; // CRC_EXTRA for HEARTBEAT
    const cs = mavCRC(&crc_input);
    // Assemble
    buf[0] = MAVLINK_STX;
    @memcpy(buf[1..6],   &hdr);
    @memcpy(buf[6..15],  &p);
    buf[15] = @as(u8, @truncate(cs & 0xFF));
    buf[16] = @as(u8, @truncate(cs >> 8));
}

// ---------------------------------------------------------------------------
// UART write stub
// ---------------------------------------------------------------------------
fn uartWrite(data: []const u8) void { _ = data; }

// ---------------------------------------------------------------------------
// Delay stub (SysTick in real firmware)
// NOTE: This stub approximates 84 cycles/µs at 84 MHz (STM32F4).
// Compiler optimisations and pipeline effects make raw nop-counts unreliable.
// Replace with SysTick-based delay or hardware timer CCR compare in
// production firmware.
// ---------------------------------------------------------------------------
fn delayUs(us: u32) void {
    var i: u32 = 0;
    while (i < us * 84) : (i += 1) asm volatile ("nop");
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

export fn main() noreturn {
    // ESC arm
    motorSetAll(CFG.pwm_min_us);
    delayUs(2_000_000);

    // PID setup
    const rollG  = PidGains{ .kp = 1.2, .ki = 0.03, .kd = 0.15 };
    const pitchG = PidGains{ .kp = 1.2, .ki = 0.03, .kd = 0.15 };
    const yawG   = PidGains{ .kp = 0.8, .ki = 0.01, .kd = 0.00 };
    var rollS  = PidState{};
    var pitchS = PidState{};
    var yawS   = PidState{};

    // Mission
    var mission = MissionManager{};
    _ = mission.addWaypoint(.{ .lat = 47.39, .lon = 8.54, .alt_m = 30 });
    _ = mission.addWaypoint(.{ .lat = 47.40, .lon = 8.54, .alt_m = 30, .release_payload = true });
    mission.start();

    // Safety
    var safety = SafetyMonitor{};

    var hb_buf: [17]u8 = undefined;
    var seq: u8    = 0;
    var hb_tick: u32 = 0;

    while (true) {
        // Sensor fusion
        var raw: [CFG.num_imu_instances]ImuSample = undefined;
        for (0..CFG.num_imu_instances) |i| raw[i] = readIMUInstance(@intCast(i));
        const imu = voteIMU(raw);

        // RC input
        var rc = readRC();
        safety.tick(rc);
        if (safety.needsRTL()) {
            rc.throttle = 500.0;
            rc.roll = 0; rc.pitch = 0; rc.yaw = 0;
        }

        // Attitude PID
        const ro = rollS.update(rollG,   rc.roll,  imu.gx, DT);
        const po = pitchS.update(pitchG, rc.pitch, imu.gy, DT);
        const yo = yawS.update(yawG,     rc.yaw,   imu.gz, DT);

        // Mix + output
        const motor_us = mixHexX(rc.throttle, ro, po, yo);
        for (motor_us, 0..) |us, i| motorSet(i, us);

        // Mission progress
        if (mission.currentWaypoint()) |wp| {
            if (wp.release_payload) {
                // trigger payload GPIO
            }
        }

        // Heartbeat
        hb_tick += 1;
        if (hb_tick >= CFG.heartbeat_period) {
            hb_tick = 0;
            encodeHeartbeat(seq, &hb_buf);
            uartWrite(&hb_buf);
            seq +%= 1;
        }

        delayUs(1_000_000 / CFG.loop_hz);
    }
}
